src/wodin.ts

```diff
@@ -0,0 +1,168 @@
+     * will be the same as `names.length` and `mode.length` (i.e., one
+     * trace per series), and `y[i].length` will be the same as
+     * `x.length` (i.e., each series as one point per step)
+     */
+    y: number[][];
+    /**
+     * The mode that each trace plays. These have some predictable
+     * behaviours:
+     *
+     * * `Individual` traces will always appear before the single
+     *   `Mean` trace
+     * * For a given value in `names`, there will either be at least
+     *   one `Individual` trace, followed by exactly one `Mean` trace
+     *   or there will be a single `Deterministic` trace
+     */
+    mode: DiscreteSeriesMode[];
+}
+
+interface DiscreteSolution {
+    info: DustModelInfo;
+    state: DustStateTime;
+    times: number[];
+}
+
+export function runModelDiscrete(Model: DustModelConstructable,
+                                 pars: UserType, timeStart: number,
+                                 timeEnd: number, dt: number,
+                                 nParticles: number): DiscreteSolution {
+    const stepStart = Math.floor(timeStart / dt);
+    const stepEnd = Math.ceil(timeEnd / dt);
+    const d = new Dust(Model, pars, nParticles, stepStart);
+    const info = d.info();
+
+    const steps = seq(stepStart, stepEnd); // inclusive
+    const times = steps.map((s) => s * dt);
+    const state = d.simulate(steps, null);
+    return { info, state, times };
+}
+
+/**
+ * Run a discrete time model for wodin.  The basic idea here is that
+ * we will run some number of replicate simulations forward in time
+ * and immediately summarise these in a way that Wodin can use.
+ *
+ * There is some sleight of hand here with time that currently copies
+ * the behaviour from the shiny app. The discrete time models don't
+ * really have a concept of time, they just have "steps", and there is
+ * a time variable (`time` in most of the short course/MSc models)
+ * which is defined as `time = step * dt`. So if we know what `dt` is
+ * then we can work directly in terms of "time" in the interface and
+ * obscure steps a bit. So here you need to provide `timeStart`,
+ * `timeEnd` and `dt` from which we can arrange a suitable range of
+ * steps to run the model, then return things in terms of this scaled
+ * time.  For example, if `dt` was 0.1 and you ran the model from time
+ * 0 to 50 we would return 501 points (steps 0, 1, ..., 499, 500)
+ * corresponding to times 0.0, 0.1, 49.9, 50.0. If `dt` is not the
+ * inverse of a natural number this will behave poorly but practically
+ * that's not a big deal because the course organisers bake dt into
+ * the fixed code.
+ *
+ * @param Model The model constructor
+ *
+ * @param pars Parameters to set into the model on construction
+ *
+ * @param timeStart The starting step (often 0)
+ *
+ * @param timeEnd The finishing step, greater than zero
+ *
+ * @param dt The size of each step
+ *
+ * @param nParticles The number of independent particles to run
```

Comment:
This should be the number of variables?

Reply:
this is the number of replicate runs, it's unrelated to the number of variables

Reply:
ah, of course!